// @nghacks/image-viewer/bundles/nghacks-image-viewer.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/dialog'), require('@angular/common'), require('rxjs'), require('rxjs/operators'), require('@angular/material/progress-spinner'), require('@angular/material/button'), require('@angular/material/icon')) :
    typeof define === 'function' && define.amd ? define('@nghacks/image-viewer', ['exports', '@angular/core', '@angular/material/dialog', '@angular/common', 'rxjs', 'rxjs/operators', '@angular/material/progress-spinner', '@angular/material/button', '@angular/material/icon'], factory) :
    (global = global || self, factory((global.nghacks = global.nghacks || {}, global.nghacks['image-viewer'] = {}), global.ng.core, global.ng.material.dialog, global.ng.common, global.rxjs, global.rxjs.operators, global.ng.material.progressSpinner, global.ng.material.button, global.ng.material.icon));
}(this, (function (exports, core, dialog, common, rxjs, operators, progressSpinner, button, icon) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                __createBinding(exports, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ImageDimension = /** @class */ (function () {
        function ImageDimension(width, height) {
            if (width === void 0) { width = 0; }
            if (height === void 0) { height = 0; }
            this.height = height;
            this.width = width;
        }
        return ImageDimension;
    }());
    var ImageViewerComponent = /** @class */ (function () {
        function ImageViewerComponent(_nativeDocument, _data, _platformId, _dialogRef) {
            this._nativeDocument = _nativeDocument;
            this._data = _data;
            this._platformId = _platformId;
            this._dialogRef = _dialogRef;
            this._unsubscribeAll = new rxjs.Subject();
            this._document = this._nativeDocument;
        }
        ImageViewerComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                var _this = this;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _a = this;
                            return [4 /*yield*/, this.getOriginalImageDimension(this._data.src)];
                        case 1:
                            _a._originalImageDimension = _b.sent();
                            this.setImageDimension();
                            this.src = this._data.src;
                            this.alt = this._data.alt;
                            this.imageLoaded = true;
                            // hide the close button before dialog close
                            // Otherwise it apears inside image
                            this._dialogRef.beforeClosed()
                                .pipe(operators.takeUntil(this._unsubscribeAll))
                                .subscribe(function (res) {
                                _this.hideCloseBtn = true;
                            });
                            return [2 /*return*/];
                    }
                });
            });
        };
        ImageViewerComponent.prototype.setImageDimension = function () {
            if (!common.isPlatformBrowser(this._platformId)) {
                return;
            }
            var windowDimension = new ImageDimension(this.percentage(this._document.body.clientWidth, 90), this.percentage(this._document.body.clientHeight, 90));
            this.calculatedImageDimension = this.calculateAspectRatioFit(this._originalImageDimension, windowDimension);
        };
        ImageViewerComponent.prototype.calculateAspectRatioFit = function (srcDimension, maxDimension) {
            if (srcDimension.height <= maxDimension.height && srcDimension.width <= maxDimension.width) {
                return srcDimension;
            }
            var ratio = Math.min(maxDimension.width / srcDimension.width, maxDimension.height / srcDimension.height);
            return new ImageDimension(srcDimension.width * ratio, srcDimension.height * ratio);
        };
        ImageViewerComponent.prototype.getOriginalImageDimension = function (imageUrl) {
            if (!imageUrl) {
                return;
            }
            return new Promise(function (res) {
                var img = new Image();
                img.onload = function (event) {
                    var loadedImage = event.currentTarget;
                    var dimension = new ImageDimension(loadedImage.width, loadedImage.height);
                    return res(dimension);
                };
                img.src = imageUrl;
            });
        };
        ImageViewerComponent.prototype.percentage = function (total, percent) {
            return ((percent / 100) * total);
        };
        ImageViewerComponent.prototype.ngOnDestroy = function () {
            this._unsubscribeAll.next();
            this._unsubscribeAll.complete();
        };
        return ImageViewerComponent;
    }());
    ImageViewerComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'lib-image-viewer',
                    template: "<figure *ngIf=\"imageLoaded; else loaderTemplate\">\n\n  <img\n    [src]=\"src\"\n    [alt]=\"alt\"\n    [style.width]=\"calculatedImageDimension?.width+'px'\"\n    [style.height]=\"calculatedImageDimension?.height+'px'\"\n  >\n</figure>\n\n<div class=\"image-view-close-button\" *ngIf=\"!hideCloseBtn\">\n  <button mat-icon-button mat-dialog-close>\n    <mat-icon>close</mat-icon>\n  </button>\n</div>\n\n\n<ng-template #loaderTemplate>\n  <div class=\"img-loading-spinner-container\">\n    <mat-spinner></mat-spinner>\n  </div>\n</ng-template>\n\n\n",
                    styles: ["figure{margin:0;min-height:72px;min-width:72px;padding:0}.img-loading-spinner-container{align-items:center;display:flex;height:150px;justify-content:center;width:150px}.image-view-close-button{animation:fade-in .2s linear both;animation-delay:.2s;cursor:pointer;position:fixed!important;right:-100px;top:-100px}.image-view-close-button button{background-color:rgba(0,0,0,.4);color:#fff;height:72px;width:72px}@keyframes fade-in{0%{right:-100px;top:-100px}to{right:-20px;top:-20px}}::ng-deep .ng-hacks-image-viewer{max-width:none!important}::ng-deep .ng-hacks-image-viewer .mat-dialog-container{background:transparent;border-radius:0;box-shadow:none;overflow:hidden;padding:0}::ng-deep .ng-hacks-image-viewer-backdrop{background:rgba(0,0,0,.32);cursor:zoom-out}"]
                },] }
    ];
    ImageViewerComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
        { type: undefined, decorators: [{ type: core.Inject, args: [dialog.MAT_DIALOG_DATA,] }] },
        { type: undefined, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
        { type: dialog.MatDialogRef }
    ]; };
    ImageViewerComponent.propDecorators = {
        setImageDimension: [{ type: core.HostListener, args: ['window:resize', ['$event'],] }]
    };

    var ImageViewerDirective = /** @class */ (function () {
        function ImageViewerDirective(_elRef, _dialog) {
            this._elRef = _elRef;
            this._dialog = _dialog;
            if (this._elRef.nativeElement.tagName === 'IMG') {
                this._cursor = 'zoom-in';
            }
        }
        Object.defineProperty(ImageViewerDirective.prototype, "disabled", {
            /**
             * @description Disables image viewer
             */
            set: function (v) {
                this._disabled = v;
                if (v) {
                    this._cursor = 'initial';
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(ImageViewerDirective.prototype, "customSrc", {
            set: function (v) {
                if (!v) {
                    return;
                }
                this._customSrc = v;
            },
            enumerable: false,
            configurable: true
        });
        ImageViewerDirective.prototype.viewImage = function () {
            if (this._disabled) {
                return;
            }
            this._dialog.open(ImageViewerComponent, {
                autoFocus: false,
                data: {
                    src: this._customSrc || this.src,
                    alt: this.alt
                },
                panelClass: 'ng-hacks-image-viewer',
                backdropClass: 'ng-hacks-image-viewer-backdrop'
            });
        };
        return ImageViewerDirective;
    }());
    ImageViewerDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[imageViewer]'
                },] }
    ];
    ImageViewerDirective.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: dialog.MatDialog }
    ]; };
    ImageViewerDirective.propDecorators = {
        _cursor: [{ type: core.HostBinding, args: ['style.cursor',] }],
        src: [{ type: core.Input }],
        alt: [{ type: core.Input }],
        disabled: [{ type: core.Input, args: ['disabled',] }],
        customSrc: [{ type: core.Input, args: ['imageViewer',] }],
        viewImage: [{ type: core.HostListener, args: ['click',] }]
    };

    var ImageViewerModule = /** @class */ (function () {
        function ImageViewerModule() {
        }
        return ImageViewerModule;
    }());
    ImageViewerModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        ImageViewerComponent,
                        ImageViewerDirective
                    ],
                    imports: [
                        common.CommonModule,
                        dialog.MatDialogModule,
                        progressSpinner.MatProgressSpinnerModule,
                        button.MatButtonModule,
                        icon.MatIconModule
                    ],
                    exports: [
                        ImageViewerDirective
                    ]
                },] }
    ];

    /*
     * Public API Surface of image-viewer
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ImageViewerDirective = ImageViewerDirective;
    exports.ImageViewerModule = ImageViewerModule;
    exports.ɵa = ImageViewerComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
